Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeSwitcher.cpp

```diff
@@ -0,0 +1,381 @@
+        if (!newlySelectedEntityIds.empty() && newlySelectedEntityIds.size() == 1)
+        {
+            AZ::ComponentApplicationBus::BroadcastResult(
+                entity, &AZ::ComponentApplicationBus::Events::FindEntity, AZ::EntityId(newlySelectedEntityIds.front()));
+
+            if (entity)
+            {
+                EntityIdList selectedEntityIds;
+                ToolsApplicationRequestBus::BroadcastResult(selectedEntityIds, &ToolsApplicationRequestBus::Events::GetSelectedEntities);
+        
+                if (selectedEntityIds.size() > 1)
+                {
+                    RemoveExclusiveComponents(*entity);
+                    return;
+                }
+
+                for (const auto& entityComponent : entity->GetComponents())
+                {
+                    // Find if the component has a component mode using GetNumOfEventHandlers
+                    const auto entityComponentIdPair = AZ::EntityComponentIdPair(entity->GetId(), entityComponent->GetId());
+                    AddComponentButton(entityComponentIdPair);
+                }
+            }
+        }
+        else if (!newlyDeselectedEntityIds.empty())
+        {
+        /*    AZStd::vector<AZ::EntityComponentIdPair> pairIdsToRemove;
+            for (auto componentDataIt : m_addedComponents)
+            {
+                pairIdsToRemove.push_back(componentDataIt.m_pairId);
+            }
+            for (auto pairId : pairIdsToRemove)
+            {
+                RemoveComponentButton(pairId);
+            }*/
+            //v.erase(std::remove_if(v.begin(), v.end(), is_odd), v.end());
+            for (AZStd::vector<ComponentData>::reverse_iterator it = m_addedComponents.rbegin(); it != m_addedComponents.rend(); ++it)
+            {
+                RemoveComponentButton(it->m_pairId);
+            }
+        }
+    }
+
+    void ComponentModeSwitcher::AddComponentButton(const AZ::EntityComponentIdPair pairId)
+    {
+        // Check if the component has a component mode.
+        const auto numHandlers = AzToolsFramework::ComponentModeFramework::ComponentModeDelegateRequestBus::GetNumOfEventHandlers(pairId);
+        if (numHandlers < 1)
+        {
+            return;
+        }
+
+        ComponentData newComponentData = ComponentData(pairId);
+
+        // If the component has not already been added as a button, add the button
+        if (auto componentDataIt = AZStd::find_if(
+                m_addedComponents.begin(),
+                m_addedComponents.end(),
+                [newComponentData](const ComponentData& componentInfo)
+                {
+                    return componentInfo.m_component->RTTI_GetType() == newComponentData.m_component->RTTI_GetType();
+                });
+            componentDataIt == m_addedComponents.end())
+        {
+            AddSwitcherButton(newComponentData, newComponentData.m_componentName.c_str(), newComponentData.m_iconPath.c_str());
```

Comment:
update: As we're passing through `newComponentData` and it already has the component name and icon path, we could just remove the second two parameters here (unless this function is used in another context somewhere else and these need to be separate?)

Reply:
Nice catch, I made the function before I had the component data struct and didn't think about getting rid of the old inputs

Reply:
Nice one thanks!